rotkehlchen/tasks/assets.py

```diff
@@ -381,3 +391,57 @@ def update_aave_v3_underlying_assets(chains_aggregator: 'ChainsAggregator') -> N
             name=DBCacheStatic.LAST_AAVE_V3_ASSETS_UPDATE,
             value=ts_now(),
         )
+
+
+def maybe_detect_new_tokens(database: 'DBHandler') -> None:
+    if not CachedSettings().get_settings().auto_detect_tokens:
+        return
+
+    with database.conn.read_ctx() as cursor:
+        tracked_accounts = {address_tuple[0] for address_tuple in cursor.execute(
+            'SELECT DISTINCT account FROM blockchain_accounts;',
+        )}
+
+        last_save_time = database.get_last_balance_save_time(cursor)
+        detected_tokens = defaultdict(list)
+        # we query the decoded history events that are the first event of each distinct
+        # unignored asset for that account that happened on or after last_save_time.
+        for event_data in cursor.execute(
+            # events that are the earliest events of distinct assets after last_save_time
+            f'SELECT {HISTORY_BASE_ENTRY_FIELDS}, '
+            f'{EVM_EVENT_FIELDS} {EVM_EVENT_JOIN} {filter_ignore_asset_query()} '
+            'AND timestamp >= ? GROUP BY asset, location_label;',
+            (ts_sec_to_ms(last_save_time),),
+        ):
+            event = EvmEvent.deserialize_from_db(event_data[1:])
+            if (
+                event.location not in EVM_LOCATIONS or
+                not event.asset.is_evm_token() or
+                event.maybe_get_direction() != EventDirection.IN
+            ):
+                # if the location or asset is not evm, or if event is not IN, we skip it
```

Comment:
no need for the comment. It's obvious

```suggestion
            ):
```